LUCENE-9856: fail precommit on unused local variables

rmuir opened the pull request:
See the JIRA for more details.

This turns on ecj option: `ecj.javadocs.prefs:org.eclipse.jdt.core.compiler.problem.unusedLocal=error`
For problematic generated code, mark with `SuppressWarnings("unused")`. 

For this to really prevent the linter from failing, also set `org.eclipse.jdt.core.compiler.problem.suppressOptionalErrors=enabled`

Set both options for eclipse IDE config as well.

Fix `gradle regenerate` to work again with modern python, some code was importing `gcd` from the wrong place (at least for python 3.9).

mikemccand committed:
LUCENE-9856: enable ecj unused local variable detection; run 'gradle ecjLint' to see all problems

mikemccand committed:
LUCENE-9856: fix a few unused local variables

rmuir committed:
LUCENE-9856: remove more unused local variables

This fixes all the instances, except for the ones in generated code.

rmuir committed:
LUCENE-9856: add module-specific overrides

I am sure it can be done more "groovy" and needs to be rewritten, but it works.

rmuir committed:
Merge branch 'main' into jira/LUCENE-9856-static-analysis

rmuir commented:
This change is largely mechanical: simply remove the unused local variables. Try to preserve any method calls that may have side effects (e.g. throw exception) in tests.

In general, it is usually a pretty bad smell to have such unused variables: the idea is not to go sniffing around that and debug deeply, but just to get the stuff fixed so the linter can stop new stuff from coming in.

If you want to help fix the issues (or make the gradle better!!!), please, feel free to just push commits to the branch!

rmuir committed:
Revert "LUCENE-9856: add module-specific overrides"

This reverts commit 66afbd682492e5209047dbe37afb138d9dadd479.

rmuir committed:
LUCENE-9856: allow SuppressWarnings("unused") to disable unusedLocal check via secret ecj option!

Still need to fix queryparser, but this fixes expressions

rmuir commented:
OK i reverted the gradle changes completely, i found a secret option so that `SuppressWarnings` can be used. This fixes the expressions module, but we need to apply `SuppressWarnings("unused")` to some of the queryparser generated code. I am looking into it.

rmuir committed:
LUCENE-9856: fix gradlew regenerate (broken for py 3.9), fix javacc regeneration to suppress warnings

Also add the changed config items to the eclipse IDE config.

rmuir committed:
LUCENE-9856: remove dead code (unusedInstance)

Also update ecj to latest for the latest checks.

The new check is not yet enabled, need to address generated code. But it
is very productive...

rmuir committed:
LUCENE-9856: enable unusedPrivateMember and suppress(unused) more gen'd code

Including that monster DFA.

rmuir committed:
LUCENE-9856: enable/fix more unnecessary code detection in linter

rmuir committed:
LUCENE-9856: override equals = must override hashcode too, @deprecated must @Deprecated too

Easy ones to enforce since nothing fails.

rmuir commented:
ok, I think ecj linter is now in better shape, especially as far as dead code goes.

But wow, the amount of dead `private` members and methods that we had was really surprising.

We should followup by enabling the check to enforce `@Override` annotations, so that when public/protected methods get removed, dangling bits aren't left around. That's a good way to improve it to the future. But, another issue. I am tired.

zacharymorn commented:
Thanks Robert for working on this! And yeah I didn't quite anticipate so much dead code lingering around when I created the issue! From your latest comment this PR is very much ready now? Any aspect I can still look into in this PR to help out as well at this point? (and sorry for coming in late in the discussion!)

In addition, I have a newbie question. It appears error prone was integrated fairly recently this year https://issues.apache.org/jira/browse/LUCENE-9497, but the issue itself didn't seems to mention why error prone was needed when there's already ecj lint, and which one should be preferred for which scenario. Is the plan to have both ecj & error prone to co-exist in Lucene to catch different kinds of bugs?

dweiss reviewed (approved):
LGTM.

dweiss commented:
> Is the plan to have both ecj & error prone to co-exist in Lucene to catch different kinds of bugs?

I think the union of all these tools is different than their intersection so they help catch issues, even if they do some duplicate work?

uschindler commented on gradle/generation/javacc.gradle:
```diff
@@ -95,6 +95,12 @@ def commonCleanups = { FileTree generatedFiles ->
           "// (setDebugStream omitted).")
+      text = text.replace(
+          "public class QueryParserTokenManager ",
+          "@SuppressWarnings(\"unused\") public class QueryParserTokenManager ")
```
Hi, To get rid of the stupid escapes for quotes, use single quotes around:

```
'@SuppressWarnings("unused") public class QueryParserTokenManager'
```

IMHO I like single quoes in Groovy more, as they don't do expansion of properties/variables inside. I always use single quotes in Grooy to prevent expansion :-)

rmuir replied:
I knew this would come up: but don't see single quotes being used much in our gradle build.. is this really consensus? 

Why do we run spotless across the java code not but the groovy code too which is an even crazier language?

rmuir replied:
I opened https://issues.apache.org/jira/browse/LUCENE-9863 to fix this

uschindler replied:
That's fine. I generally try to avoid double quotes in groovy, because I hate the case that it expands variables inside. Personally, I see languages expanding variables inside strings as a security risk.

rmuir replied:
I will fix it here, but we must wait a bit. This change requires regenerating the huge DFA :(

rmuir replied:
And i had to close a bunch of apps and restart it already :)

`Mar 23 07:59:25 kernel: Out of memory: Killed process 410597 (java) total-vm:15959272kB, anon-rss:11760000kB, file-rss:0kB, shmem-rss:32kB, UID:1000 pgtables:23356kB oom_score_adj:0`

uschindler commented on lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/ConcatenateGraphFilter.java:
```diff
@@ -348,7 +348,7 @@ private static Automaton replaceSep(Automaton a, Character tokenSeparator) {
    */
   public static final class BytesRefBuilderTermAttributeImpl extends AttributeImpl
-      implements BytesRefBuilderTermAttribute, TermToBytesRefAttribute {
+      implements BytesRefBuilderTermAttribute {
```
LOL

uschindler commented on lucene/analysis/common/src/test/org/apache/lucene/analysis/cjk/TestCJKBigramFilterFactory.java:
```diff
@@ -56,11 +56,10 @@ public void testHanOnlyUnigrams() throws Exception {
 
   /** Test that bogus arguments result in exception */
   public void testBogusArguments() throws Exception {
-    IllegalArgumentException expected =
```
I think those are indeed unused, but the reason why they are there is a different one: This allows step-by-step debugging. By declaring an unused variable, you can step through the code / set breakpoints and see what the variable contains. Without it, most debuggers don't show the return value (depend on your IDE).

I agree to remove them, but this may annoy people.

rmuir replied:
I am pretty sure I added this test and I don't use debuggers. If someone wants unused, they can add `SuppressWarnings("unused")` to their test.

uschindler replied:
This was just a general comment about "why do people this?". If it was you, no problem. It just looked like "I want debugging, so I assign everything to a local variable".

uschindler replied:
So I just mentioned it here, so people that may complain are aware that we "noticed" that behaviour.

rmuir replied:
I honestly don't think there is a lot of the behavior going on. Maybe because IDEs tend to issue warnings when you do this? 

We fixed only about 140 violations for this check total across the whole codebase including tests.

uschindler replied:
Obviously it comes from copy'n'paste = shit'n'waste. We have many places in code where the exception's message is checked. Code like this one here does not do it. Because this syntax with the lambda for exception checking was "new" at this point in time lot of people shit'n'wasted. 😱

uschindler commented on lucene/analysis/common/src/test/org/apache/lucene/analysis/core/TestAnalyzers.java:
```diff
@@ -109,7 +109,9 @@ public void _testStandardConstants() {
     x = StandardTokenizer.HIRAGANA;
     x = StandardTokenizer.KATAKANA;
     x = StandardTokenizer.HANGUL;
+    assert x >= 0;
```
Should't this be a real assertTrue() of JUnit?

rmuir replied:
the entire test is crazy, recommend looking at it in context. It is a "compile-time" "test" to ensure the constants are "public". We can remove the asserts, now that `SuppressWarnings("unused")` does the right thing

uschindler commented on lucene/core/src/test/org/apache/lucene/search/TestTotalHits.java:
```diff
@@ -22,6 +22,7 @@
 
 public class TestTotalHits extends LuceneTestCase {
 
+  @SuppressWarnings("unlikely-arg-type")
```
Just for interest: what does this complain about?

rmuir replied:
This one detects `equals(typeA, typeB)` where typeA is different  than typeB.

In `src/java` every single thing it found was a bug (fixed in this PR). Stuff like code calling equals(BytesRefBuilder, BytesRef), which probably was broken during a refactoring.

In `src/test` it also found similar bugs. But there are a few tests that intentionally do something like `assertFalse(myObj.equals("a string or something"))` to test their `equals` method on a different class.

only a few tests needed the suppression, so I just added it to those equals tests.

uschindler commented on lucene/core/src/test/org/apache/lucene/store/TestMultiMMap.java:
```diff
@@ -181,7 +181,7 @@ public void testSeekZero() throws Exception {
   public void testSeekSliceZero() throws Exception {
     int upto = TEST_NIGHTLY ? 31 : 3;
-    for (int i = 0; i < 3; i++) {
+    for (int i = 0; i < upto; i++) {
```
woah, thanks for fixing!

uschindler reviewed (approved):
Looks fine to me. I left some comments, mostly informational or as an optional improvement, no need to change anything.

rmuir committed:
Merge branch 'main' into jira/LUCENE-9856-static-analysis

rmuir commented:
> Thanks Robert for working on this! And yeah I didn't quite anticipate so much dead code lingering around when I created the issue! From your latest comment this PR is very much ready now? Any aspect I can still look into in this PR to help out as well at this point? (and sorry for coming in late in the discussion!)
> 

We may always followup with more checks. I mentioned a good one in my comment above, i'd really like to enforce that `@Override` annotation is used everywhere, to help prevent dead `public` and `protected` methods in the future? 

In the preferences file, we would set `org.eclipse.jdt.core.compiler.problem.missingOverrideAnnotation=error` and then fix all the problems. It would be good as a standalone followup issue

> In addition, I have a newbie question. It appears error prone was integrated fairly recently this year https://issues.apache.org/jira/browse/LUCENE-9497, but the issue itself didn't seems to mention why error prone was needed when there's already ecj lint, and which one should be preferred for which scenario. Is the plan to have both ecj & error prone to co-exist in Lucene to catch different kinds of bugs?

I don't know the error-prone well, but it is more of a "static analysis tool": I immediately think that it leans towards heuristics and "false positives". 

On the other hand, ECJ is a "compiler", but with a lot of nice checks. Maybe they are not as fancy as the kind of checks that error-prone might be able to do, instead they are simpler, more like "additional compiler warnings". We've used ECJ for a long time: It runs efficiently, has a TON more eyes looking at its output every day, and provides great error messages. IMO, for solving your problem of "my IDE can detect it is dead code, why doesn't the build", it is the perfect tool: since it is the thing doing that in your IDE (to my knowledge not just eclipse but intellij may use it too).

uschindler commented:
> On the other hand, ECJ is a "compiler", but with a lot of nice checks. Maybe they are not as fancy as the kind of checks that error-prone might be able to do, instead they are simpler, more like "additional compiler warnings". We've used ECJ for a long time: It runs efficiently, has a TON more eyes looking at its output every day, and provides great error messages. IMO, for solving your problem of "my IDE can detect it is dead code, why doesn't the build", it is the perfect tool: since it is the thing doing that in your IDE (to my knowledge not just eclipse but intellij may use it too).

I fully agree with that. I was also thinking about further improving compile time in Java by using ECJ as the official compiler and move it from "Lint" stage. We can maybe reimplement compileJava tasks in Gradle to use ECJ (maybe it works out of box!?). So people would see those errors as soon as they compile or run tests. ECJ is just an alternative compiler and may work as replacement for javac.

This is no suggestion, but something to discuss on a separate issue.

rmuir commented:
I was also pleasantly surprised to see how well maintained this compiler is. It is nice to see things like JDK16 support, as opposed to error-prone which reaches into java internals and seemingly has no plan for JDK16 support at all! Also, they seem to pay attention to the performance especially, as it impacts a lot of IDE users. You can see this from the commits: https://github.com/eclipse/eclipse.jdt.core

rmuir committed:
LUCENE-9856: minor tweaks

TestAnalyzers tests doesn't need asserts, it had
SuppressWarnings("unused") the whole time, so we can revert the changes
since this is now respected.

Gradle has single quotes instead of double quotes, 30 minutes later :)

mikemccand reviewed (approved):
This looks awesome!  Thank you @rmuir for driving forward on this so crazy quickly!  It is so wild that this began life as a small spinoff from another PR (https://github.com/apache/lucene/pull/25).

uschindler commented:
When you fixed the stuff, maybe reopen a new pull request with same branch?

rmuir commented:
I still have the branch locally. I think the fix is to revert this commit: https://github.com/apache/lucene/pull/34/commits/7d823c84e3f7b6bebd25eda76f9e482016c78868

Redundant superinterface was too aggressive on my part. I was trying to enable checks around "unnecessary code" and this was one of them, but I think it borked the attributes? Seems it should be deferred for now, the gain isn't worth the pain :)

We may still enable the `unusedLabel` check from that commit, I think it is safe.

zacharymorn commented:
> > Is the plan to have both ecj & error prone to co-exist in Lucene to catch different kinds of bugs?
> 
> I think the union of all these tools is different than their intersection so they help catch issues, even if they do some duplicate work?


> > Thanks Robert for working on this! And yeah I didn't quite anticipate so much dead code lingering around when I created the issue! From your latest comment this PR is very much ready now? Any aspect I can still look into in this PR to help out as well at this point? (and sorry for coming in late in the discussion!)
> 
> We may always followup with more checks. I mentioned a good one in my comment above, i'd really like to enforce that `@Override` annotation is used everywhere, to help prevent dead `public` and `protected` methods in the future?
> 
> In the preferences file, we would set `org.eclipse.jdt.core.compiler.problem.missingOverrideAnnotation=error` and then fix all the problems. It would be good as a standalone followup issue
> 
> > In addition, I have a newbie question. It appears error prone was integrated fairly recently this year https://issues.apache.org/jira/browse/LUCENE-9497, but the issue itself didn't seems to mention why error prone was needed when there's already ecj lint, and which one should be preferred for which scenario. Is the plan to have both ecj & error prone to co-exist in Lucene to catch different kinds of bugs?
> 
> I don't know the error-prone well, but it is more of a "static analysis tool": I immediately think that it leans towards heuristics and "false positives".
> 
> On the other hand, ECJ is a "compiler", but with a lot of nice checks. Maybe they are not as fancy as the kind of checks that error-prone might be able to do, instead they are simpler, more like "additional compiler warnings". We've used ECJ for a long time: It runs efficiently, has a TON more eyes looking at its output every day, and provides great error messages. IMO, for solving your problem of "my IDE can detect it is dead code, why doesn't the build", it is the perfect tool: since it is the thing doing that in your IDE (to my knowledge not just eclipse but intellij may use it too).



> > On the other hand, ECJ is a "compiler", but with a lot of nice checks. Maybe they are not as fancy as the kind of checks that error-prone might be able to do, instead they are simpler, more like "additional compiler warnings". We've used ECJ for a long time: It runs efficiently, has a TON more eyes looking at its output every day, and provides great error messages. IMO, for solving your problem of "my IDE can detect it is dead code, why doesn't the build", it is the perfect tool: since it is the thing doing that in your IDE (to my knowledge not just eclipse but intellij may use it too).
> 
> I fully agree with that. I was also thinking about further improving compile time in Java by using ECJ as the official compiler and move it from "Lint" stage. We can maybe reimplement compileJava tasks in Gradle to use ECJ (maybe it works out of box!?). So people would see those errors as soon as they compile or run tests. ECJ is just an alternative compiler and may work as replacement for javac.
> 
> This is no suggestion, but something to discuss on a separate issue. There's also a Grade plugin to do this: https://github.com/TwoStone/gradle-eclipse-compiler-plugin (and it allows to supply the properties file with warning/error config).


Thanks for the explanation and insight on ecj vs. error prone. They make a lot of sense! To keep the fun going, I have opened another Jira to enforce the use of `@Override` with `org.eclipse.jdt.core.compiler.problem.missingOverrideAnnotation=error` suggested, and will try to see if I can get a PR out soon.
